File: oc/按钮控制图片/按钮控制图片/MZView.m
//
//  MZView.m
//  按钮控制图片
//
//  Created by Mr.Yang on 16/9/19.
//  Copyright © 2016年 MZ. All rights reserved.
//

#import "MZView.h"

#define phoneHeigh [UIScreen mainScreen].bounds.size.height
#define phoneWidth [UIScreen mainScreen].bounds.size.width
#define yeCol [UIColor colorWithRed:222/255.0 green:171/255.0 blue:72/255.0 alpha:1]

#define everyHeigh 5.0f
@implementation MZView{
    
    int leftCenter;
    int rightCenter;
    
    int touHeigh;
    int zhongHeigh;
    
}
- (instancetype)init {
    self = [super init];
    if (self) {
        leftCenter = phoneHeigh / 3.0;
        rightCenter = phoneHeigh / 3.0;
        
        touHeigh = 0;
        zhongHeigh = 0;
        
        self.backgroundColor = yeCol;
        self.frame = [UIScreen mainScreen].bounds;
        
        for (int i = 0; i < 4; i++) {
            UIButton *butt1 = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
            butt1.tag = i + 100;
            butt1.backgroundColor = [UIColor greenColor];
            [self addSubview:butt1];
            butt1.center = CGPointMake((0 ==i % 2) ? phoneWidth / 2.0 - 25 : phoneWidth / 2.0 + 25, phoneHeigh / 4 * 3.0 + 50 * (i / 2));
            [butt1 addTarget:self action:@selector(changeImage:) forControlEvents:UIControlEventTouchUpInside];
        }
    }
    return self;
}

//改变图形
- (void)changeImage:(UIButton *)but {
    switch (but.tag) {
        case 100:{
            touHeigh += 1;
            if (2 == touHeigh) {
                touHeigh = 1;
            } else {
                leftCenter = phoneHeigh / 3.0 - (touHeigh * everyHeigh);
                [self setNeedsDisplay];
            }
        }
            break;
        case 101:{
            zhongHeigh += 1;
            if (2 == zhongHeigh) {
                zhongHeigh = 1;
            } else {
                rightCenter = phoneHeigh / 3.0 - (zhongHeigh * everyHeigh);
                [self setNeedsDisplay];
            }
        }
            break;
        case 102:{
            touHeigh -= 1;
            if (-2 == touHeigh) {
                touHeigh = -1;
            } else {
                leftCenter = phoneHeigh / 3.0 - (touHeigh * everyHeigh);
                [self setNeedsDisplay];
            }
        }
            break;
        case 103:{
            zhongHeigh -= 1;
            if (-2 == zhongHeigh) {
                zhongHeigh = -1;
            } else {
                rightCenter = phoneHeigh / 3.0 - (zhongHeigh * everyHeigh);
                [self setNeedsDisplay];
            }
        }
            break;
            
        default:
            break;
    }
    NSLog(@"dasdsadsa");
}
-(void)drawRect:(CGRect)rect {
//    CGFloat lineWidth = 25.0f;
//    
//    CGColorRef color1 = [UIColor redColor].CGColor;
//    CGContextRef ctx1 = UIGraphicsGetCurrentContext();
//    CGContextSetStrokeColorWithColor(ctx1, color1);
//    CGContextSetLineWidth(ctx1, lineWidth);
//    CGContextMoveToPoint(ctx1, phoneWidth / 4.0, leftCenter);//开始的中间坐标
//    CGContextAddLineToPoint(ctx1, phoneWidth / 2.0, rightCenter);//结束的中间坐标
//    CGContextStrokePath(ctx1);
//    
//    
//    CGColorRef color = [UIColor redColor].CGColor;
//    CGContextRef ctx = UIGraphicsGetCurrentContext();
//    CGContextSetStrokeColorWithColor(ctx, color);
//    CGContextSetLineWidth(ctx, lineWidth);
//    CGContextMoveToPoint(ctx, phoneWidth / 2.0, rightCenter);//开始的中间坐标
//    CGContextAddLineToPoint(ctx, phoneWidth * 3.0 / 4.0, phoneHeigh / 3.0);//结束的中间坐标
//    CGContextStrokePath(ctx);
    //1.创建UIBezierPath

    [self drawSiBianTopLeft:CGPointMake(phoneWidth/4.0, leftCenter - 15)
                   topRight:CGPointMake(phoneWidth/2.0, rightCenter - 15)
                 bottomLeft:CGPointMake(phoneWidth/4.0, leftCenter + 15)
                bottomRight:CGPointMake(phoneWidth/2.0, rightCenter + 15)
                     isleft:1];
    
    [self drawSiBianTopLeft:CGPointMake(phoneWidth/2.0, rightCenter - 15)
                   topRight:CGPointMake(phoneWidth/4.0 * 3, phoneHeigh / 3.0 - 15)
                 bottomLeft:CGPointMake(phoneWidth/2.0, rightCenter + 15)
                bottomRight:CGPointMake(phoneWidth/4.0 * 3, phoneHeigh / 3.0 + 15)
                     isleft:0];
    
    //中间虚线
    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(phoneWidth/2.0, rightCenter - 15)];
    [path addLineToPoint:CGPointMake(phoneWidth/2.0, rightCenter + 15)];
    
    CGFloat dash[] = {4,4};
    
    /**
     *  设置路径线路的图案
     *
     *  @param pattern 线段和间隙的长度???  C类型数组
     *  @param count   图案的值??? pattern中数据个数
     *  @param phase   在该偏移开始绘制图案，在沿虚线图案点测量??? 起始位置
     */
    
    [path setLineDash:dash count:2 phase:0];
    [yeCol setStroke];
    [path stroke];
    
    //边线
    
}

- (void)drawSiBianTopLeft:(CGPoint)topLeftPoint topRight:(CGPoint)topRightPoint bottomLeft:(CGPoint)bottomLeftPoint bottomRight:(CGPoint)bottomRightPoint isleft:(BOOL)left{
    //1.创建UIBezierPath
    UIBezierPath *path = [UIBezierPath bezierPath];
    //2。设计路径
    [path moveToPoint:topLeftPoint];
    [path addLineToPoint:topRightPoint];
    [path addLineToPoint:bottomRightPoint];
    [path addLineToPoint:bottomLeftPoint];
    [path addLineToPoint:topLeftPoint];
    [path addLineToPoint:CGPointMake((topLeftPoint.x + topRightPoint.x) / 2.0, (topLeftPoint.y + topRightPoint.y) / 2.0)];
    //路径的其他设置
    path.lineWidth = 30;//粗细
    path.lineCapStyle = kCGLineCapRound;//线头样式
    path.lineJoinStyle = kCGLineJoinRound; //设置线头相交的部分
    //3.设计描边,填充
    [[UIColor whiteColor] setStroke];
    [[UIColor whiteColor] setFill];
    //4.绘制
    [path stroke];
    [path fill];
}






























@end

File: swift/SwiftTest/SwiftTest/MZView.swift
//
//  MZView.swift
//  SwiftTest
//
//  Created by Mr.Yang on 16/9/21.
//  Copyright © 2016年 MZ. All rights reserved.
//

import UIKit

let  phoneHeigh = UIScreen.main.bounds.size.height
let  phoneWidth = UIScreen.main.bounds.size.width
let  yeCol = UIColor.init(red: 222/255.0, green: 171/255.0, blue: 72/255.0, alpha: 1);
let  everyHeigh  = 5.0;
class MZView: UIView {
    
    var leftCenter = phoneHeigh / 3.0;
    var rightCenter = phoneHeigh / 3.0;
    
    var touHeigh = 0.0;
    var zhongHeigh = 0.0;
    
    override init(frame: CGRect) {
        super.init(frame: frame);
        self.normalCreateButton();
    }
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder);
        self.normalCreateButton();
    }
    
    func normalCreateButton() {
        self.frame = UIScreen.main.bounds;
        self.backgroundColor = yeCol;
        
        for i in 0..<4 {
            let butt = UIButton.init();
            butt.frame = CGRect.init(x: 0, y: 0, width: 20, height: 20)
            butt.center = CGPoint.init(x: (0 == i % 2) ? phoneWidth / 2 - 25 : phoneWidth / 2 + 25, y: phoneHeigh / 4 * 3 + 50.0 * CGFloat(i / 2));
            butt.tag = i + 100;
            butt.backgroundColor = UIColor.green;
            self.addSubview(butt);
            butt.addTarget(self, action: #selector(changeImage(butt:)), for: UIControlEvents.touchUpInside)
        }

    }
    
    func changeImage(butt : UIButton) {
        switch butt.tag {
        case 100:
            self.touHeigh += 1;
            if (2 == self.touHeigh) {
                self.touHeigh = 1;
            } else {
                leftCenter = phoneHeigh / 3.0 - CGFloat(touHeigh * everyHeigh);
                self.setNeedsDisplay();
            }
        
        case 101:
            self.zhongHeigh += 1;
            if (2 == self.zhongHeigh) {
                self.zhongHeigh = 1;
            } else {
                rightCenter = phoneHeigh / 3.0 - CGFloat(zhongHeigh * everyHeigh);
                self.setNeedsDisplay();
            }
            
        case 102:
            self.touHeigh -= 1;
            if (-2 == self.touHeigh) {
                self.touHeigh = -1;
            } else {
                leftCenter = phoneHeigh / 3.0 - CGFloat(touHeigh * everyHeigh);
                self.setNeedsDisplay();
            }
            
        case 103:
            self.zhongHeigh -= 1;
            if (-2 == self.zhongHeigh) {
                self.zhongHeigh = -1;
            } else {
                rightCenter = phoneHeigh / 3.0 - CGFloat(zhongHeigh * everyHeigh);
                self.setNeedsDisplay();
            }

        default: break
            
        }
    }
    
    override func draw(_ rect: CGRect) {
        self.drawSiBianTopLeft(topLeftPoint: CGPoint.init(x: phoneWidth/4.0, y: leftCenter - 15),
                               topRightPoint: CGPoint.init(x: phoneWidth/2.0, y: rightCenter - 15),
                               bottomLeftPoint: CGPoint.init(x: phoneWidth/4.0, y: leftCenter + 15),
                               bottomRightPoint: CGPoint.init(x: phoneWidth/2.0, y: rightCenter + 15),
                               left: true);
        
        self.drawSiBianTopLeft(topLeftPoint: CGPoint.init(x: phoneWidth/2.0, y: rightCenter - 15),
                               topRightPoint: CGPoint.init(x: phoneWidth/4.0 * 3, y: phoneHeigh / 3.0 - 15),
                               bottomLeftPoint: CGPoint.init(x: phoneWidth/2.0, y: rightCenter + 15),
                               bottomRightPoint: CGPoint.init(x: phoneWidth/4.0 * 3, y: phoneHeigh / 3.0 + 15),
                               left: false);
        let path = UIBezierPath.init();
        path.move(to: CGPoint.init(x: phoneWidth/2.0, y: rightCenter - 15));
        path.addLine(to: CGPoint.init(x: phoneWidth/2.0, y: rightCenter + 15));
        
        let dash:[CGFloat] = [4,4];
        
        path.setLineDash(dash, count: 2, phase: 0)
        yeCol.setStroke();
        path.stroke();
        
    }
    func drawSiBianTopLeft(topLeftPoint:CGPoint, topRightPoint:CGPoint, bottomLeftPoint:CGPoint, bottomRightPoint:CGPoint, left:Bool) {
        let path = UIBezierPath.init();
        
        path.move(to: topLeftPoint);
        path.addLine(to: topRightPoint);
        path.addLine(to: bottomRightPoint);
        path.addLine(to: bottomLeftPoint);
        path.addLine(to: topLeftPoint);
        path.addLine(to: CGPoint.init(x: (topLeftPoint.x + topRightPoint.x) / 2.0, y: (topLeftPoint.y + topRightPoint.y) / 2.0));
        
        path.lineWidth = 30;
        path.lineCapStyle = CGLineCap(rawValue: Int32(1))!;
        path.lineJoinStyle = CGLineJoin(rawValue: Int32(1))!;
        
        UIColor.white.setStroke();
        UIColor.white.setFill();
        
        path.stroke();
        path.fill();
    }
}





























